torch/nn/attention/templated_attention.py

```diff
@@ -0,0 +1,170 @@
+from typing import Callable
```

Comment:
hmm maybe this file should go under `torch/_higher_order_ops`? Right now everything in this file is technically public API, right 🤔

Reply:
Makes sense, I do ultimately want the public api to live here, but can move the implementation details to higher_order_ops/